README
WINDOWS USERS: 
Be sure to set your PATH to [install_path]/q3map3/binaries where 'install_path' is the 
git repository's root folder. Binaries contains all of the .dll files necessary for 
GTKRadiant.exe and q3map3.exe to run. 

*NIX USERS: 
Follow the installation directions using scons. 


****NOTE****
I've renamed the STLport folder to STLportl as a quick way to compile without
Visual Studio complaining about pre-existing functions.

Rename the STLportl file back to STLport if you want to use it. 

mhash-0.9/NEWS
mhash ChangeLog ($Date: 2004/04/17 21:36:08 $)
===============================================================================

People

   NM - Nikos Mavroyanopoulos <nmav@hellug.gr>
   SS - Sascha Schumann <sascha@schumann.cx>

Version 0.9.1
- Added RIPEMD128/256/320, SHA224/384/512, Whirlpool
- Really, this time!
  (jkmhash@itaparica.org)

Version 0.8.19
- Added VC++ project files, contributed by James M. Rogers 
  <james@singlestep.com>

Version 0.8.18 (released March 6 2003)
- Several additions and fixes to built win32 dlls.

Version 0.8.17 (released October 4 2002)
- Corrected the HEX key mode. Refuses to convert hex strings longer
  than the given key size.
- Corrected some memory leaks in keygen_s2k. Patch by Bob Mathews 
  <bobmathews@alumni.calpoly.edu>
- Corrected wrong use of sizeof in hmac_test. Reported by Ellis Pritchard
  <ellispritchard@users.sf.net>

Version 0.8.16 (released May 29 2002)
- Code cleanups and optimizations
- Added configure time options to disable certain algorithms

Version 0.8.15 (released May 24 2002)
- Corrected Iterated S2K
- Made all static tables constant (pointed out by Darryl Miles <darryl@netbauds.net>)

Version 0.8.14 (release March 28 2002)
- mhash_deinit() and mhash_hmac_deinit() now accept a null digest parameter
- Corrected memory leaks (patch by Gustavo Niemeyer <niemeyer@conectiva.com>)
- Corrected bug in Gost hash algorithm (pointed out by Mike Gorchak <mike@malva.ua>)

Version 0.8.13 (released November 18 2001)
- Added ADLER32 algorithm (implementation by Manuel Kasper <mk@neon1.net>)

Version 0.8.12 (released October 29 2001)
- Corrected bugs in mhash_get_name() functions
- Added SHA-256 from libnettle.

Version 0.8.11 (released October 21 2001)
- added mhash_save_state_mem() and mhash_restore_state_mem() functions,
  initial patches and idea by Blake Stephen <Stephen.Blake@veritect.com>
- Added mhash_get_name_static()
- Corrected mhash_get_name() and mhash_count()

Version 0.8.10 (released on July 12 2001)
- New mhash_deinit() and mhash_hmac_deinit() functions based on the
  proposal, and patches by William Ahern <wahern@25thandClement.com>

Version 0.8.9 (released on February 04 2001)
- TIGER runs on alpha (NM)
- Documentation updated

Version 0.8.8 (released on January 29 2001)
- MD4 and MD5 now compile on big endian machines (NM)
- TIGER now compiles on big endian machines. (NM)

Version 0.8.6 (released on January 21 2001)
- Corrected Tiger implementation (NM, bug pointed out by Gordon Mohr <gojomo@usa.net>)
- Added MD4 (NM)
- Added TIGER128 and TIGER160 (NM)

Version 0.8.5 (released on January 2001)
- added checks for malloc failures (NM, some patches by Sami Farin <sfarin@ratol.fi>)
- corrected old style checks for mhash failures (NM, reported by Joe Hohertz <jhohertz@wiretap.net>)

Version 0.8.4 (released on December 15 2000)
- added mhash_hmac_end_m() and mhash_end_m() functions which allow to specify the malloc function. (NM)
- added mhash_cp() - Copy a context (NM)
- added mhash_get_mhash_algo() - Returns the name of the algorithm of the context (NM)
- fixes in crc32b and gost (for multithreaded applications) (NM)

Version 0.8.3 (released on October 25 2000)
- Compiles under Win32 using cygwin (with -mno-cygwin) (NM)
- Bugfixes in KEYGEN_ASIS (NM)

Version 0.8.2 (released on May 12 2000)
- Corrected Memory leaks in HMAC implementation (NM, contributed by Marios Hadjieleftheriou <marioh@cs.ucr.edu>)
- Added HAVAL224 (NM)

Version 0.8.1 (released on April 11 2000)
- Bugfixes in hmac implementation. (NM)
- Removed MD2 (which was not free) (NM)

Version 0.8.0 (released on April 5 2000)
- No longer needs posix thread locks to be thread safe (NM)
- Supports unlimited number of MHASH threads (NM)
- Improved the key generation algorithm API (NM)
- Corrected Haval implementation. (NM)
  Now mhash uses Paulo Barreto's implementation
  Added HAVAL256, HAVAL192, HAVAL160, HAVAL128 (3 passes haval)
- License was changed to GNU Lesser GPL

Version 0.7.0 (released on Mar 3rd 2000)
- cleaned up binary interface (SS)
- improved POSIX Threads detection (SS)
- added keygen API (NM)
- added new improved RIPEMD160 implementation (NM)
- removed RIPEMD128 (NM)
- added RPM .spec file by work@scripty.com (SS)

Version 0.6.1 (released on Oct 23th 1999)
- fixed Tiger algorithm on 32-bit architectures (NM)

Version 0.6.0 (released on Oct 11th 1999)
- export only symbols of the public interface (SS)
- MHASH_API_VERSION has been changed to 19991004 (SS)
- API change: All algorithm names have been prefixed with MHASH_
  to avoid namespace pollution (SS)
- added mhash_get_hash_pblock() by NM, reimplemented by SS
- improved data structure handling (NM)
- use "--silent" libtool option (SS)
- added hmac functions (SS, contributed by NM)

Version 0.5.2 (released on Jul 14th 1999)
- fix VPATH build (SS)
- upgrade to libtool 1.3.3 (SS)
- rework directory layout (simplified) (SS)
- change test script to work with Bourne shells (SS)
- fix some big endian issues in RIPEMD160, RIPEMD128 and TIGER (SS)
	
Version 0.5 (relased on May 22nd 1999)
- put under CVS control
- change get_block_size to mhash_get_block_size (NM)
- add --disable-pthreads option (NM)
- add libtool and automake support (SS)
- add several algorithms (SS)
  HAVAL, RIPEMD128, RIPEMD160, TIGER, GOST, CRC32B
- add mhash_get_hash_name (SS)
- add mhash_count (SS)
- documentation revamp (SS)
- renew API (SS)
- add test vectors (SS)


mhash-0.9/lib/md4.c
/*
 *    Copyright (C) 2001 Nikos Mavroyanopoulos
 *
 *    This library is free software; you can redistribute it and/or modify it 
 *    under the terms of the GNU Library General Public License as published 
 *    by the Free Software Foundation; either version 2 of the License, or 
 *    (at your option) any later version.
 *
 *    This library is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Library General Public License for more details.
 *
 *    You should have received a copy of the GNU Library General Public
 *    License along with this library; if not, write to the
 *    Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 *    Boston, MA 02111-1307, USA.
 */


#include <libdefs.h>

#ifdef ENABLE_MD4

/* 
 * The algorithm is due to Ron Rivest.  This code is based on code
 * written by Colin Plumb in 1993.
 */
 

/*
 * This code implements the MD4 message-digest algorithm.
 */

#include "mhash_md4.h"

#ifndef WORDS_BIGENDIAN
#define byteReverse(buf, len)	/* Nothing */
#else
static void byteReverse(unsigned char *buf, unsigned longs);


/*
 * Note: this code is harmless on little-endian machines.
 */
static void byteReverse(unsigned char *buf, unsigned longs)
{
	word32 t;
	do {
		t = (word32) ((unsigned) buf[3] << 8 | buf[2]) << 16 |
		    ((unsigned) buf[1] << 8 | buf[0]);
		*(word32 *) buf = t;
		buf += 4;
	} while (--longs);
}
#endif

#define rotl32(x,n)   (((x) << ((word32)(n))) | ((x) >> (32 - (word32)(n))))

/*
 * Start MD4 accumulation.  Set bit count to 0 and buffer to mysterious
 * initialization constants.
 */
void MD4Init(struct MD4Context *ctx)
{
	ctx->buf[0] = 0x67452301;
	ctx->buf[1] = 0xefcdab89;
	ctx->buf[2] = 0x98badcfe;
	ctx->buf[3] = 0x10325476;

	ctx->bits[0] = 0;
	ctx->bits[1] = 0;
}

/*
 * Update context to reflect the concatenation of another buffer full
 * of bytes.
 */
void MD4Update(struct MD4Context *ctx, unsigned char const *buf,
	       unsigned len)
{
	register word32 t;

	/* Update bitcount */

	t = ctx->bits[0];
	if ((ctx->bits[0] = t + ((word32) len << 3)) < t)
		ctx->bits[1]++;	/* Carry from low to high */
	ctx->bits[1] += len >> 29;

	t = (t >> 3) & 0x3f;	/* Bytes already in shsInfo->data */

	/* Handle any leading odd-sized chunks */

	if (t) {
		unsigned char *p = (unsigned char *) ctx->in + t;

		t = 64 - t;
		if (len < t) {
			memcpy(p, buf, len);
			return;
		}
		memcpy(p, buf, t);
		byteReverse(ctx->in, 16);
		MD4Transform(ctx->buf, (word32 *) ctx->in);
		buf += t;
		len -= t;
	}
	/* Process data in 64-byte chunks */

	while (len >= 64) {
		memcpy(ctx->in, buf, 64);
		byteReverse(ctx->in, 16);
		MD4Transform(ctx->buf, (word32 *) ctx->in);
		buf += 64;
		len -= 64;
	}

	/* Handle any remaining bytes of data. */

	memcpy(ctx->in, buf, len);
}

/*
 * Final wrapup - pad to 64-byte boundary with the bit pattern 
 * 1 0* (64-bit count of bits processed, MSB-first)
 */
void MD4Final(struct MD4Context *ctx, unsigned char* digest)
{
	unsigned int count;
	unsigned char *p;

	/* Compute number of bytes mod 64 */
	count = (ctx->bits[0] >> 3) & 0x3F;

	/* Set the first char of padding to 0x80.  This is safe since there is
	   always at least one byte free */
	p = ctx->in + count;
	*p++ = 0x80;

	/* Bytes of padding needed to make 64 bytes */
	count = 64 - 1 - count;

	/* Pad out to 56 mod 64 */
	if (count < 8) {
		/* Two lots of padding:  Pad the first block to 64 bytes */
		memset(p, 0, count);
		byteReverse(ctx->in, 16);
		MD4Transform(ctx->buf, (word32 *) ctx->in);

		/* Now fill the next block with 56 bytes */
		memset(ctx->in, 0, 56);
	} else {
		/* Pad block to 56 bytes */
		memset(p, 0, count - 8);
	}
	byteReverse(ctx->in, 14);

	/* Append length in bits and transform */
	((word32 *) ctx->in)[14] = ctx->bits[0];
	((word32 *) ctx->in)[15] = ctx->bits[1];

	MD4Transform(ctx->buf, (word32 *) ctx->in);
	byteReverse((unsigned char *) ctx->buf, 4);
	
	if (digest!=NULL)
		memcpy(digest, ctx->buf, 16);
	memset(ctx, 0, sizeof(ctx));	/* In case it's sensitive */
}

/* The three core functions */

#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (y)) | ((x) & (z)) | ((y) & (z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))

#define FF(a, b, c, d, x, s) { \
    (a) += F ((b), (c), (d)) + (x); \
    (a) = rotl32 ((a), (s)); \
  }
#define GG(a, b, c, d, x, s) { \
    (a) += G ((b), (c), (d)) + (x) + (word32)0x5a827999; \
    (a) = rotl32 ((a), (s)); \
  }
#define HH(a, b, c, d, x, s) { \
    (a) += H ((b), (c), (d)) + (x) + (word32)0x6ed9eba1; \
    (a) = rotl32 ((a), (s)); \
  }


/*
 * The core of the MD4 algorithm
 */
void MD4Transform(word32 buf[4], word32 const in[16])
{
	register word32 a, b, c, d;

	a = buf[0];
	b = buf[1];
	c = buf[2];
	d = buf[3];

	FF(a, b, c, d, in[0], 3);	/* 1 */
	FF(d, a, b, c, in[1], 7);	/* 2 */
	FF(c, d, a, b, in[2], 11);	/* 3 */
	FF(b, c, d, a, in[3], 19);	/* 4 */
	FF(a, b, c, d, in[4], 3);	/* 5 */
	FF(d, a, b, c, in[5], 7);	/* 6 */
	FF(c, d, a, b, in[6], 11);	/* 7 */
	FF(b, c, d, a, in[7], 19);	/* 8 */
	FF(a, b, c, d, in[8], 3);	/* 9 */
	FF(d, a, b, c, in[9], 7);	/* 10 */
	FF(c, d, a, b, in[10], 11);	/* 11 */
	FF(b, c, d, a, in[11], 19);	/* 12 */
	FF(a, b, c, d, in[12], 3);	/* 13 */
	FF(d, a, b, c, in[13], 7);	/* 14 */
	FF(c, d, a, b, in[14], 11);	/* 15 */
	FF(b, c, d, a, in[15], 19);	/* 16 */

	GG(a, b, c, d, in[0], 3);	/* 17 */
	GG(d, a, b, c, in[4], 5);	/* 18 */
	GG(c, d, a, b, in[8], 9);	/* 19 */
	GG(b, c, d, a, in[12], 13);	/* 20 */
	GG(a, b, c, d, in[1], 3);	/* 21 */
	GG(d, a, b, c, in[5], 5);	/* 22 */
	GG(c, d, a, b, in[9], 9);	/* 23 */
	GG(b, c, d, a, in[13], 13);	/* 24 */
	GG(a, b, c, d, in[2], 3);	/* 25 */
	GG(d, a, b, c, in[6], 5);	/* 26 */
	GG(c, d, a, b, in[10], 9);	/* 27 */
	GG(b, c, d, a, in[14], 13);	/* 28 */
	GG(a, b, c, d, in[3], 3);	/* 29 */
	GG(d, a, b, c, in[7], 5);	/* 30 */
	GG(c, d, a, b, in[11], 9);	/* 31 */
	GG(b, c, d, a, in[15], 13);	/* 32 */

	HH(a, b, c, d, in[0], 3);	/* 33 */
	HH(d, a, b, c, in[8], 9);	/* 34 */
	HH(c, d, a, b, in[4], 11);	/* 35 */
	HH(b, c, d, a, in[12], 15);	/* 36 */
	HH(a, b, c, d, in[2], 3);	/* 37 */
	HH(d, a, b, c, in[10], 9);	/* 38 */
	HH(c, d, a, b, in[6], 11);	/* 39 */
	HH(b, c, d, a, in[14], 15);	/* 40 */
	HH(a, b, c, d, in[1], 3);	/* 41 */
	HH(d, a, b, c, in[9], 9);	/* 42 */
	HH(c, d, a, b, in[5], 11);	/* 43 */
	HH(b, c, d, a, in[13], 15);	/* 44 */
	HH(a, b, c, d, in[3], 3);	/* 45 */
	HH(d, a, b, c, in[11], 9);	/* 46 */
	HH(c, d, a, b, in[7], 11);	/* 47 */
	HH(b, c, d, a, in[15], 15);	/* 48 */


	buf[0] += a;
	buf[1] += b;
	buf[2] += c;
	buf[3] += d;
}

#endif /* ENABLE_MD4 */
